site/docs/skylark/windows_tips.md

```diff
@@ -0,0 +1,264 @@
+---
+layout: documentation
+title: Writing rules on Windows
+---
+
+# Writing rules on Windows
+
+Common problems of writing platform-independent rules and ideas to overcome them.
+
+This document focuses on writing rules that work on Windows.
+
+## Paths
+
+Problems:
+
+- **Length limit**: maximum path length is 259 characters.
+
+  While Windows also supports paths up to 32767 characters ("long paths"), many programs are built
+  with the lower limits. The action your rule executes may not handle long paths either.
+
+- **Working directory**: is also limited to 259 characters.
+
+  Processes cannot `cd` into a directory longer than 259 characters.
+
+- **Case-sensitivity**: Windows paths are case-insensitive, Unix paths are case-sensitive.
+
+  Be aware of this when creating command lines for actions.
+
+- **Path separators**: are backslash (`\`), not forward slash (`/`).
+
+  Bazel uses paths with `/` separators, so on Windows you have to replace those with `\` when
+  creating comamnd lines and environment variables for actions.
```

Comment:
In many cases `/` works on Windows, from what I remember. Do you know more about it?

Reply:
Indeed. `/` support is inconsistent.

It works when the binary you run supports it. (Typically those binaries use the ANSI subset of the Windows API. They are also the ones that don't support long paths.)

Built-in cmd.exe commands (like `copy`, `del`) tend not to support it:
```
C:\foo>copy bar/baz/qux bar/baz/qox
The syntax of the command is incorrect.

C:\foo>copy bar\baz\qux bar/baz/qox
The syntax of the command is incorrect.

C:\foo>copy bar\baz\qux bar\baz\qox
        1 file(s) copied.
```

but sometimes they do (e.g. when piping):

```
C:\foo>copy bar\baz\qux bar\baz\qox
        1 file(s) copied.

C:\foo>echo x > bar/baz/qaax

C:\foo>type bar/baz/qaax
The syntax of the command is incorrect.

C:\foo>type bar\baz\qaax
x
```

Reply:
Maybe put one line about it.